src/realm/db_element.hpp

```diff
@@ -0,0 +1,686 @@
+/*************************************************************************
+ *
+ * Copyright 2016 Realm Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ **************************************************************************/
+
+#ifndef REALM_DB_ELEMENT_HPP
+#define REALM_DB_ELEMENT_HPP
+
+#include <realm/alloc.hpp>
+
+namespace realm {
+
+// Maximum number of bytes that the payload of an DbElement can be
+const size_t max_array_payload = 0x00ffffffL;
+
+/// Special index value. It has various meanings depending on
+/// context. It is returned by some search functions to indicate 'not
+/// found'. It is similar in function to std::string::npos.
+const size_t npos = size_t(-1);
+
+/// Alias for realm::npos.
+const size_t not_found = npos;
+
+struct TreeInsertBase {
+    size_t m_split_offset;
+    size_t m_split_size;
+};
+
+class ArrayParent {
+public:
+    virtual ~ArrayParent() noexcept
+    {
+    }
+
+protected:
+    virtual void update_child_ref(size_t child_ndx, ref_type new_ref) = 0;
+
+    virtual ref_type get_child_ref(size_t child_ndx) const noexcept = 0;
+
+    // Used only by Array::to_dot().
+    virtual std::pair<ref_type, size_t> get_to_dot_parent(size_t ndx_in_parent) const = 0;
+
+    friend class DbElement;
+    friend class Array;
+};
+
+class DbElement {
```

Comment:
I'm not sure I understand what this name is supposed to represent. Could you add some documentation to clarify the use of this class? (I know you posted a 2 line summary in the PR, but that will be lost as soon as it is merged).
Might I also suggest using `DatabaseElement` instead of `DbElement`. Acronyms are the root of all evil :P. So is CamelCase.